src/main/java/org/jenkinsci/plugins/workflow/cps/CpsBodyExecution.java

```diff
@@ -172,29 +180,86 @@ private Continuable createContinuable(CpsThread currentThread, CpsCallableInvoca
     }
 
     @Override
-    public synchronized Collection<StepExecution> getCurrentExecutions() {
-        if (thread==null)   return Collections.emptyList();
-
-        StepExecution s = thread.getStep();
-        if (s!=null)        return Collections.singleton(s);
-        else                return Collections.emptyList();
+    public Collection<StepExecution> getCurrentExecutions() {
+        CpsThread t;
+        synchronized (this) {
+            t = thread;
+            if (t == null) {
+                return Collections.emptySet();
+            }
+        }
+        final SettableFuture<Collection<StepExecution>> result = SettableFuture.create();
+        t.getExecution().runInCpsVmThread(new FutureCallback<CpsThreadGroup>() {
+            @Override public void onSuccess(CpsThreadGroup g) {
+                try {
+                    List<StepExecution> executions = new ArrayList<>();
+                    // cf. trick in CpsFlowExecution.getCurrentExecutions(true)
+                    Map<FlowHead, CpsThread> m = new LinkedHashMap<>();
+                    for (CpsThread t : g.threads.values()) {
+                        m.put(t.head, t);
+                    }
+                    for (CpsThread t : m.values()) {
+                        // TODO seems cumbersome to have to go through the flow graph to find out whether a head is a descendant of ours, yet FlowHead does not seem to retain a parent field
+                        LinearBlockHoppingScanner scanner = new LinearBlockHoppingScanner();
```

Comment:
It's a little more terse and does a prefilter to only return block start nodes by instanceof check (maybe slightly faster).  Not a big deal though, there's a half dozen ways to do this, I just want to put it on your radar since this is exactly what it's for.